splinterd/src/config/partial.rs

```diff
@@ -0,0 +1,338 @@
+/// several sources. As such, all values of the PartialConfig are options as it is not necessary
+/// to provide all values from a single source.
+#[derive(Deserialize, Default, Debug)]
+pub struct PartialConfig {
+    storage: Option<String>,
+    transport: Option<String>,
+    cert_dir: Option<String>,
+    ca_certs: Option<String>,
+    client_cert: Option<String>,
+    client_key: Option<String>,
+    server_cert: Option<String>,
+    server_key: Option<String>,
+    service_endpoint: Option<String>,
+    network_endpoint: Option<String>,
+    peers: Option<Vec<String>>,
+    node_id: Option<String>,
+    bind: Option<String>,
+    #[cfg(feature = "database")]
+    database: Option<String>,
+    registry_backend: Option<String>,
+    registry_file: Option<String>,
+    heartbeat_interval: Option<u64>,
+}
+
+impl PartialConfig {
+    /// Creates a new PartialConfig object from a toml file. Available to use when the `config-toml`
+    /// feature flag is not in use.
+    #[cfg(not(feature = "config-toml"))]
+    pub fn from_file(mut f: File) -> Result<PartialConfig, ConfigError> {
+        let mut toml = String::new();
+        f.read_to_string(&mut toml)?;
+
+        toml::from_str::<PartialConfig>(&toml).map_err(ConfigError::from)
+    }
+
+    pub fn storage(&self) -> Option<String> {
+        self.storage.clone()
+    }
+
+    pub fn transport(&self) -> Option<String> {
+        self.transport.clone()
+    }
+
+    pub fn cert_dir(&self) -> Option<String> {
+        self.cert_dir.clone()
+    }
+
+    pub fn ca_certs(&self) -> Option<String> {
+        self.ca_certs.clone()
+    }
+
+    pub fn client_cert(&self) -> Option<String> {
+        self.client_cert.clone()
+    }
+
+    pub fn client_key(&self) -> Option<String> {
+        self.client_key.clone()
+    }
+
+    pub fn server_cert(&self) -> Option<String> {
+        self.server_cert.clone()
+    }
+
+    pub fn server_key(&self) -> Option<String> {
+        self.server_key.clone()
+    }
+
+    pub fn service_endpoint(&self) -> Option<String> {
+        self.service_endpoint.clone()
+    }
+
+    pub fn network_endpoint(&self) -> Option<String> {
+        self.network_endpoint.clone()
+    }
+
+    pub fn peers(&self) -> Option<Vec<String>> {
+        self.peers.clone()
+    }
+
+    pub fn node_id(&self) -> Option<String> {
+        self.node_id.clone()
+    }
+
+    pub fn bind(&self) -> Option<String> {
+        self.bind.clone()
+    }
+
+    #[cfg(feature = "database")]
+    pub fn database(&self) -> Option<String> {
+        self.database.clone()
+    }
+
+    pub fn registry_backend(&self) -> Option<String> {
+        self.registry_backend.clone()
+    }
+
+    pub fn registry_file(&self) -> Option<String> {
+        self.registry_file.clone()
+    }
+
+    pub fn heartbeat_interval(&self) -> Option<u64> {
+        self.heartbeat_interval
+    }
+
+    #[cfg(feature = "config-builder")]
```

Comment:
Is attributing each method with cfg an established style? I think it would be cleaner to wrap all of the functions for each feature in a block and then just apply the attribute to the block.

Reply:
The disadvantage of that is that if you remove the cfg, then it would change the intent (larger future diff). So the current approach is actually good.

I assume these are here because otherwise we get compiler warnings.

Reply:
The compiler warnings assumption is correct.